### ISSUE TITLE
(uf-ho) Fatal failure at src/theory/theory_model.cpp:672

### ISSUE THREAD

**Issue by muchang** [cvc5/cvc5#5371]
For this formula, CVC4 throws out a fatal failure. It is a regression from cvc4-1.8.
```
[542] % cvc4-1.8 --uf-ho small.smt2
sat
[543] % 
[543] % cvc4 --uf-ho small.smt2
Fatal failure within void CVC4::theory::TheoryModel::assignFunctionDefinition(CVC4::Node, CVC4::Node) at src/theory/theory_model.cpp:672
Check failure
 f_def.isConst()
Non-constant f_def: (lambda ((BOUND_VARIABLE_349 Bool)) (or (not BOUND_VARIABLE_349) BOUND_VARIABLE_349))
Aborted
[544] % 
[544] % cat small.smt2
(set-logic ALL)
(declare-fun a (Bool) Bool)
(assert (a false))
(assert (a true))
(check-sat)
[545] %
```
OS: Ubuntu 18.04
Commit: 42fe4ae

**Comment by muchang** [comment]
Another case:
```
[553] % cvc4 --strings-exp --uf-ho -q small.smt2
Fatal failure within void CVC4::theory::TheoryModel::assignFunctionDefinition(CVC4::Node, CVC4::Node) at /local/suz-local/software/CVC4/src/theory/theory_model.cpp:665
Check failure

 d_uf_models.find(f) == d_uf_models.end()

Aborted
[554] % 
[554] % cat small.smt2
(declare-fun a (Bool) Bool)
(declare-fun b (Bool) Bool)
(assert (forall ((c Bool)) (= (a c) (b c))))
(check-sat)
[555] % 
```

**Comment by HanielB** [comment]
This issue no longer occurs. Probably fixed by #6697, which improved the interplay between the HOL setup and the rest of the solver (particularly with `--strings-exp`).

**Comment by rainoftime** [comment]
at c8bc4881d4f3bf542
~~~~
(set-option :uf-ho true)
(set-option :finite-model-find true)
(declare-sort s 0)
(declare-fun k (s s) Bool)
(declare-fun r (s s) Bool)
(assert (= true (forall ((? s)) (or (not (k ? ?)) (not (r ? ?))))))
(check-sat)
~~~~
~~~~
cvc5 -q xx.smt2
Fatal failure within void cvc5::theory::TheoryModel::assignFunctionDefinition(cvc5::Node, cvc5::Node) at /home/cvc5/src/theory/theory_model.cpp:698
Check failure

 d_uf_models.find(f) == d_uf_models.end()

Aborted
~~~~
